libs/remote_storage/src/simulate_failures.rs

```diff
@@ -0,0 +1,123 @@
+//! This module provides a wrapper around a real RemoteStorage implementation that
+//! causes some operations to fail. For testing purposes.
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::Mutex;
+
+use crate::{Download, DownloadError, RemotePath, RemoteStorage, StorageMetadata};
+
+pub struct UnreliableWrapper {
+    inner: crate::GenericRemoteStorage,
+
+    fail_first: u64,
+
+    attempts: Mutex<HashMap<RemoteOp, u64>>,
+}
+
+#[derive(Debug, Hash, Eq, PartialEq)]
+enum RemoteOp {
+    List,
+    ListPrefixes(Option<RemotePath>),
+    Upload(RemotePath),
+    Download(RemotePath),
+    Delete(RemotePath),
+}
+
+impl UnreliableWrapper {
+    pub fn new(inner: crate::GenericRemoteStorage, fail_first: u64) -> Self {
+        assert!(fail_first > 0);
+        UnreliableWrapper {
+            inner,
+            fail_first,
+            attempts: Mutex::new(HashMap::new()),
+        }
+    }
+
+    ///
+    /// Common functionality for all operations.
+    ///
+    /// On the first attempts of this operation, return an error. After 'fail_first'
+    /// attempts, let the operation go ahead, and clear the counter.
+    ///
+    fn attempt(&self, op: RemoteOp) -> anyhow::Result<u64> {
```

Comment:
> I think that is the root cause of the discussion, and that anyhow::Error parameter in From impl comes from this anyhow::Result.

The reason I added the `From` impl is that I changed the return type of `list_prefixes` from `anyhow::Result` to `Result<, DownloadError>`. Without the From, have to change the ? calls in `list_prefixes` to `map_err`.

I'm not sure about a separate SimulatedFailure error type, but it would be useful to distingush between "retryable" and "non-retryable" errors. For example, if you use and invalid parameter, or the object doesn't exist, it's never going to succeed and there's no point in retrying it. But if the network is down, that should be retried. That would need a lot more work though.

Reply:
Yes, error type handling requries more work than `anyhow` and we're supposed to have somebody to outline the module borders and make good errors on that borders.
This PR does something opposite and that's why it feels odd.

Yet, the idea of the retryable operations seems also interesting, so feel free to resolve if you think that should be the vector, not the custom variants for test cases.

Reply:
>The reason I added the From impl is that I changed the return type of list_prefixes from anyhow::Result to Result<, DownloadError>. Without the From, have to change the ? calls in list_prefixes to map_err.

And `download_layer_file` too

Reply:
Ok, removed the From impl now, and used `map_err`s instead. This would benefit from more distinguishing retryable and non-retryable errors more carefully, but that's a nice-to-have. Doing some unnecessary retries before failing hard is not a big deal.